tests/features/layers/activation-order/test/testcase.cpp

```diff
@@ -0,0 +1,177 @@
+  }
+};
+
+TEST_F(LayerActivationOrder, BaseLayerHasNotRegressed) {
+  SingleKeyTest(Key_A);
+}
+
+TEST_F(LayerActivationOrder, ShifToLayerOne) {
+  // Pressing (3,6) shifts to Layer 1, and we stay there until release.
+
+  sim_.Press(3, 6); // ShiftToLayer(1)
+  auto state = RunCycle();
+
+  SingleKeyTest(Key_B);
+
+  // Releasing (3,6) gets us back to the base layer
+  sim_.Release(3, 6); // ShiftToLayer(1)
+  state = RunCycle();
+
+  SingleKeyTest(Key_A);
+}
+
+TEST_F(LayerActivationOrder, ShiftingWithCaching) {
+  // Pressing (0, 0) will activate A
+  sim_.Press(0, 0);
+  auto state = RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 1);
+  EXPECT_THAT(
+    state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+    Contains(Key_A));
+
+  // Pressing (3, 6) will switch to Layer 1
+  sim_.Press(3, 6);
+  state = RunCycle();
+
+  // ...since we're still pressing (0, 0), and there was no change in the HID
+  // states, we shouldn't emit a report.
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 0);
+
+  // Pressing (0, 1), the report shall contain 'A' _and_ 'B'.
+  sim_.Press(0, 1);
+  state = RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 1);
+  EXPECT_THAT(
+    state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+    Contains(Key_A));
+  EXPECT_THAT(
+    state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+    Contains(Key_B));
+
+  // Releasing (0, 0), the report should now contain B only
+  sim_.Release(0, 0);
+  state = RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 1);
+  EXPECT_THAT(
+    state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+    Contains(Key_B));
+
+  // Release (0, 1)
+  sim_.Release(0, 1);
+  state = RunCycle();
+
+  // Test B in isolation again
+  SingleKeyTest(Key_B);
+
+  // Release the layer key as well.
+  sim_.Release(3, 6);
+  state = RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 0);
+}
+
+TEST_F(LayerActivationOrder, Ordering) {
+  // Pressing (3, 9) will switch to Layer 2
+  sim_.Press(3, 9);
+  auto state = RunCycle();
+
+  // Pressing (0, 0) will activate 'C'
+  sim_.Press(0, 0);
+  state = RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 1);
+  EXPECT_THAT(
+    state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+    Contains(Key_C));
+
+  // Pressing (3, 6) will activate Layer 1
+  sim_.Press(3, 6);
+  state = RunCycle();
+
+  // Pressing (0, 1) will activate 'B' now, due to activation ordering.
+  sim_.Press(0, 1);
+  state = RunCycle();
+
+  ASSERT_EQ(state->HIDReports()->Keyboard().size(), 1);
+  EXPECT_THAT(
+    state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+    Contains(Key_C));
+  EXPECT_THAT(
+    state->HIDReports()->Keyboard(0).ActiveKeycodes(),
+    Contains(Key_B));
+
+  sim_.Release(0, 0);
+  sim_.Release(0, 1);
+  sim_.Release(3, 6);
+  sim_.Release(3, 9);
+  state = RunCycle();
+}
+
+TEST_F(LayerActivationOrder, LayerZero) {
+  sim_.Press(3, 7); // Macro#0: Layer.deactivate(0)
```

Comment:
Rather than hard-coding these numbers in the test, here, these should be a preprocessor macro with a descriptive name and the comment should be on the macro.

Reply:
How about a constexpr instead, to encourage better habits elsewhere?

Reply:
Yup, I'm going with constexpr. I'm going to touch the harness too, because `sim_.Press()` & friends currently take a pair of `uint8_t`, need to teach them to have a `KeyAddr` variant too.